sd_mecha/merge_methods.py

```diff
@@ -608,11 +680,95 @@ def overlapping_sets_pmf(n, p, overlap, overlap_emphasis):
+def model_stock_for_tensor(
+    *deltas: Tensor | LiftFlag[MergeSpace.DELTA],
+    cos_eps: Hyper = 1e-6,    
+    **kwargs,
+) -> Tensor | LiftFlag[MergeSpace.DELTA]:
+
+    # This is obvious.
+    w_avg = n_average.__wrapped__(*deltas)
+
+    # t can get inf so handle with care
+    t = get_model_stock_t(deltas, cos_eps)
+
+    # return w_h. Notice that w_0 is 0 here.
+    return torch.nan_to_num(t * w_avg)
+
+# The guess from mergekit: Average of cos(theta). Expected value is 0, somehow match with paper.
+# However this may be very unstable, and the range is still -1 to 1.
+def get_model_stock_t(deltas, cos_eps):
+    n = len(deltas)
+
+    # Generator function. Default eps from torch API doc.
+    cos = torch.nn.CosineSimilarity(dim=-1, eps=cos_eps)
+
+    # One-liner is all you need. I may make it in running average if it really memory hungry.
+    cos_thetas = [cos(deltas[i], deltas[i + 1]) for i, _ in enumerate(deltas) if (i + 1) < n]
+    
+    # Still a vector.
+    cos_theta = torch.stack(cos_thetas).mean(dim=0)
+
+    # Convert to column vector for multiplication.
+    t = (n * cos_theta / (1 + (n - 1) * cos_theta)).unsqueeze(-1)
+
+    return t
+
+# This becomes a wrapper since I want TIES use GM also.
+@convert_to_recipe
+def geometric_median(
+    *models: Tensor | SameMergeSpace,
+    eps: Hyper = 1e-6,    
+    maxiter: Hyper = 100, 
+    ftol: Hyper = 1e-20,
+    **kwargs,
+)-> Tensor | SameMergeSpace:
+    return geometric_median_list_of_array(models, eps, maxiter, ftol)
+
+# Original sourcecode: https://github.com/krishnap25/geom_median/blob/main/src/geom_median/torch/weiszfeld_list_of_array.py
+# Changed to "List comprehension" and rely on torch API only. It is now fully parallel.
+def geometric_median_list_of_array(models, eps, maxiter, ftol):
+    # I think it is impossible to pass this from user space so I hardcode this instead.
+    # Meanwhile I rename "points" as "models"
+    # no_grad part is rare case: Merge algorithm under GPU is never heard.
+    weights = torch.ones(len(models), device=models[0].device)
+
+    # initialize median estimate at mean
+    median = weighted_average(models, weights)
+    new_weights = weights
+    objective_value = geometric_median_objective(median, models, weights)
+
+    # Weiszfeld iterations
+    for _ in range(maxiter):
+        prev_obj_value = objective_value
+        denom = torch.stack([l2distance(p, median) for p in models])
+        new_weights = weights / torch.clamp(denom, min=eps) 
+        median = weighted_average(models, new_weights)
+
+        objective_value = geometric_median_objective(median, models, weights)
+        if abs(prev_obj_value - objective_value) <= ftol * objective_value:
+            break
+        
+    return weighted_average(models, new_weights)
+
+def weighted_average(points, weights):
+    # weighted_average_component is not even required.
+    return torch.sum(torch.stack([p * weights[i] for i, p in enumerate(points)]), dim=0) / weights.sum()
```

Comment:
Is it accurate if I say that this could take a lot of memory given it requires the same key of all models to be loaded at the same time?

If that is the case, then is it possible to implement the geometric median in an iterated manner? If not, then I guess there is no way around this.

Reply:
- For the [pythonic list comprehension](https://www.w3schools.com/python/python_lists_comprehension.asp), yes. Nice parallelization with $O(N)$ of memory usage. 1.216TB for 116 SDXL under fp64. Notice that 116 SDXL in fp16 is already 750GB, which is 3TB for fp64, therefore I think 40% of overall models is already memory efficient. The $O(N)$ part will be explained seperately.
- From [wiki](https://en.wikipedia.org/wiki/Geometric_median),  Weiszfeld's algorithm is already an iterlative algorithm:
> One common approach of this type, called Weiszfeld's algorithm after the work of [Endre Weiszfeld](https://en.wikipedia.org/wiki/Endre_Weiszfeld),[[16]](https://en.wikipedia.org/wiki/Geometric_median#cite_note-16) is a form of [iteratively re-weighted least squares](https://en.wikipedia.org/wiki/Iteratively_re-weighted_least_squares).
...
This method converges for almost all initial positions
- Since geometric median is NOT statistics median, it cannot be an $O(1)$ algorithm like the [n_average](https://github.com/ljleb/sd-mecha/blob/main/examples/n_average.py). A complete set of "weighted average" are required to calculate the L2 distance for all the existing models, and I believe that it is already done inplace. Strict `list.append()` approach may reduce memory footprint to $O(1)$, but it will take at least 40x time to iterlate, which will surely [exceed my implemented test case](https://github.com/6DammK9/sd-mecha/blob/main/test/unit_test_geom_median.py#L78), and my "1.216TB" merge already used 21 hours, 40x of them will spend more than 1 month to do so. Meanwhile merging a few models with this algorithm will make this algorithm sounds ineffective due to [lack of confidence](https://www.calculator.net/sample-size-calculator.html).
- For 3 models under fp16, it will probably take 30 minutes and 30GB of RAM, which is not easy either. Moving it to `cuda` doesn't help either, because it already converges and early-stopped with around 6-10 iterlations. You can try with the [test case](https://github.com/6DammK9/sd-mecha/blob/main/test/unit_test_geom_median.py#L78),  
- BTW TGMD merge (modified TIES / DARE + GeometricMedian) will not slower, and the figures above are already for TGMD. Plain GM will be 1x in time with 0.75x of memory requiement.
>  If not, then I guess there is no way around this.
- I don't know what you mean. Hope that you're not concerning about hardware requirement because most ML algorithm are $O(N^3)$ in complexity, and parameter size of SDXL model is huge (3.6b) for any algorithm, otherwise we are limited with algebra. This is still $O(N)$ like TIES / DARE / SLERP / ModelStock, not even close to rebasin, which is $O(NlogN)$ in time. 
- Also, I expect such merging is being done solely in CPU, with an average workstation, like X99 + 256GB of RAM, which should support for 100 SDXL models in fp16.

Reply:
> If not, then I guess there is no way around this.
By this I mean that the current implementation might be as optimal as we can make it to be, since there is no known way to make it take less memory while preserving the speed of execution.

Reply:
Thanks for the clarification, I see why this parallel approach was taken now.